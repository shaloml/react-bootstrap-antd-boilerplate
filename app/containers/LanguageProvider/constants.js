/*
 *
 * LanguageProvider constants
 *
 */

export const CHANGE_LOCALE = 'app/LanguageToggle/CHANGE_LOCALE';
export const LOAD_MESSAGES = 'app/LanguageToggle/LOAD_MESSAGES';
export const LOADING_MESSAGES = 'app/LanguageToggle/LOADING_MESSAGES';
