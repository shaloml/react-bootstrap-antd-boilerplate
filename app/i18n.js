/**
 * i18n.js
 *
 * This will setup the i18n language files and locale data for your app.
 *
 *
 */
import * as defaultTranslationMessages from 'translations/en.json';
import { DEFAULT_LOCALE } from './locales';

const enTranslationMessages = () => import('translations/en.json');

export const translationMessages = {
  en: enTranslationMessages
};

/**
 * Loads messages for a given locale asynchronously.  This way, the browser
 * only receives the translations that it needs.
 *
 * @param locale
 */
export const fetchMessages = (locale) =>
  translationMessages[locale]().then(({ default: messages }) =>
    formatTranslationMessages(locale, messages)
  );

export const formatTranslationMessages = (locale, messages) => {
  const defaultFormattedMessages =
    locale !== DEFAULT_LOCALE
      ? formatTranslationMessages(DEFAULT_LOCALE, defaultTranslationMessages)
      : {};
  const flattenFormattedMessages = (formattedMessages, key) => {
    const formattedMessage =
      !messages[key] && locale !== DEFAULT_LOCALE
        ? defaultFormattedMessages[key]
        : messages[key];
    return Object.assign(formattedMessages, { [key]: formattedMessage });
  };
  return Object.keys(messages).reduce(flattenFormattedMessages, {});
};

export const defaultMessages = defaultTranslationMessages;
