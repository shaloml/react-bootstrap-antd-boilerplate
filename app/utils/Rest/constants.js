export const API_NAMES = ['MAIN'];
