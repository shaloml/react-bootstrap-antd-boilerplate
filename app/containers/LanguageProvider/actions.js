/*
 *
 * LanguageProvider actions
 *
 */

import { CHANGE_LOCALE, LOAD_MESSAGES, LOADING_MESSAGES } from './constants';

export function changeLocale(languageLocale) {
  return {
    type: CHANGE_LOCALE,
    locale: languageLocale
  };
}

export function loadMessages({ locale, messages }) {
  return {
    type: LOAD_MESSAGES,
    locale,
    messages
  };
}

export function loadingMessages() {
  return {
    type: LOADING_MESSAGES
  };
}
