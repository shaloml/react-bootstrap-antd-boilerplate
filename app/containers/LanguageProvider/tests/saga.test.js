/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { CHANGE_LOCALE } from '../constants';
import { loadingMessages, loadMessages } from '../actions';

import getTranslationMessages, { getMessages } from '../saga';
import { defaultMessages } from '../../../i18n';

/* eslint-disable redux-saga/yield-effects */
describe('getTranslationMessages Saga', () => {
  let getMessagesGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getMessagesGenerator = getMessages({ type: CHANGE_LOCALE, locale: 'de' });

    const isLoading = getMessagesGenerator.next().value;
    expect(isLoading).toEqual(put(loadingMessages()));

    const callDescriptor = getMessagesGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the loadMessages action if it requests the translations successfully', () => {
    const response = {
      'this.is.a.translation': 'with a string'
    };
    const putDescriptor = getMessagesGenerator.next(response).value;
    expect(putDescriptor).toEqual(
      put(loadMessages({ locale: 'de', messages: response }))
    );
  });

  it('should call the loadMessages action with the defaultMessages if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getMessagesGenerator.throw(response).value;
    expect(putDescriptor).toEqual(
      put(loadMessages({ locale: 'en', messages: defaultMessages }))
    );
  });
});

describe('getTranslationMessagesSaga Saga', () => {
  const getTranslationMessagesSaga = getTranslationMessages();

  it('should start task to watch for CHANGE_LOCALE action', () => {
    const takeLatestDescriptor = getTranslationMessagesSaga.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(CHANGE_LOCALE, getMessages)
    );
  });
});
