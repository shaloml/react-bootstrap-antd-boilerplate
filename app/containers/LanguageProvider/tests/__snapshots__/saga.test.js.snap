// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTranslationMessages Saga should call the loadMessages action with the defaultMessages if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "de",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getTranslationMessages Saga should dispatch the loadMessages action if it requests the translations successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "de",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
